wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/LanguageServerAbstractFileWatcher.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.languageserver;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.languageserver.RegistryContainer.Registry;
+import org.eclipse.lsp4j.DidChangeWatchedFilesParams;
+import org.eclipse.lsp4j.FileChangeType;
+import org.eclipse.lsp4j.FileEvent;
+import org.eclipse.lsp4j.services.LanguageServer;
+import org.eclipse.lsp4j.services.WorkspaceService;
+
+/**
+ * Notifies registered language servers about file related events. Specifying of file event type is
+ * expected to be defined through this class inheritance.
+ *
+ * @author Dmytro Kulieshov
+ * @see <a
+ *     href="https://microsoft.github.io/language-server-protocol/specification#workspace_didChangeWatchedFiles">DidChangeWatchedFiles
+ *     Notification</a> section of LSP specification
+ */
+class LanguageServerAbstractFileWatcher implements Consumer<Path> {
+  private final EventService eventService;
+  private final Registry<Set<PathMatcher>> pathMatcherRegistry;
+  private final FileChangeType fileChangeType;
+
+  /** Language ID -> (pathMatcher -> pathConsumer) */
+  private final Map<String, Map<PathMatcher, Consumer<Path>>> eventConsumers =
+      new ConcurrentHashMap<>();
+
+  @Inject
+  LanguageServerAbstractFileWatcher(
+      EventService eventService, RegistryContainer registries, FileChangeType fileChangeType) {
+    this.eventService = eventService;
+    this.pathMatcherRegistry = registries.pathMatcherRegistry;
+    this.fileChangeType = fileChangeType;
+  }
+
+  @PostConstruct
+  protected void subscribe() {
+    eventService.subscribe(this::acceptEvent, LanguageServerInitializedEvent.class);
+  }
+
+  @PreDestroy
+  void removeAllWatchers() {
+    eventConsumers.clear();
+  }
+
+  private void acceptEvent(LanguageServerInitializedEvent event) {
+    String id = event.getId();
+    LanguageServer languageServer = event.getLanguageServer();
+    Set<PathMatcher> pathMatchers = pathMatcherRegistry.getOrDefault(id, ImmutableSet.of());
+
+    for (PathMatcher pathMatcher : pathMatchers) {
+      eventConsumers
```

Comment:
Absolutely subjective point of view. Applying a declarative approach that exactly aims to simplify the understanding of the source code, I used one of the most appropriate data structures available in the standard library - `Map`.

While you suggest more simplified way I'm confident it does not comply to object-oriented principles and breaks the data abstractions, so I will try to describe in detail the arguments that stands behind my solution.

First, let's look at the data we are dealing with: there is a set of language servers, a set of paths, and a set of event users. Each server has a set of path matcher and each path matcher has a corresponding event consumer.

The mapping between the path matcher and the event consumer can probably be represented by `Collection<Pair<PathMatcher, Consumer<Path>`, but there is no standard interface or class specifically to store pairs that I know, so I decided to use standard `Map` because it is commonly used collection that fits all requirements here and a bit shorter in symbols - `Map<PathMatcher, Consumer<Path>`.

As I said before, the `PathMatchers` to `Consumer` set (represented by `Map`) is specific for each language server, as you probably know when we need to declare a mapping for such kind of relation we again use `Map`. Mixing all those pairs in a single `Set` or `List`  as suggested will break the data abstraction that potentially may lead to errors and misunderstanding.

In general, I see that you tend to make the code shorter and more optimized to the detriment of its readability and clarity, and data structures simpler to the detriment of their correspondence with real-life entities/abstractions. This is probably a procedural, not an object-oriented and declarative approach. However I would like to emphasize that we chose Java, a powerful object-oriented language, exactly to utilize its object-oriented strengths.

Reply:
What is needed here (in order to fulfil the purpose of the class) is to remember a non-unique set of pairs of path matchers and path consumers. Where these pairs come from is completely irrelevant to the needs we have in this class. This has nothing to do with quality object-oriented design, nor declarative programming nor any of the other things you mention. My proposal is better because it expresses the intent of the code and even in doing that, it is shorter ans simpler. Maybe you can't see that.
I'm giving you good advice based on almost 30 years of professional programming where I had the opportunity to work with some of the best programmers in the world. If you don't want to listen, that's your prerogative.

Reply:
Actually I'm grateful for your advices, they develop my cognitive processes and I can assure you that I pay great attention to your remarks. However when I take decisions I try to apply the experience from different sources, including but not limited from yours. Thank you.